Vector Embeddings: Add example w/ Cloud SDK for AI (#1914)
cap-js/docs: guides/databases-hana.md

### Vector Embeddings { #vector-embeddings }

Vector embeddings let you add semantic search, recommendations, and generative AI features to your CAP application on SAP HANA Cloud. Embeddings are numeric arrays that represent the meaning of unstructured data (text, images, etc.), making it possible to compare and search for items that are semantically related to each other or a user query.

> 

Author: ```suggestion

#### Choose an Embedding Model

Choose an embedding model that fits your use case and data (for example english or multilingual text). The model determines the number of dimensions of the resulting output vector.

Use the [SAP Generative AI Hub](https://community.sap.com/t5/technology-blogs-by-sap/how-sap-s-generative-ai-hub-facilitates-embedded-trustworthy-and-reliable/ba-p/13596153) for unified consumption of embedding models and LLMs across different vendors and open source models. Check for available models on the [SAP AI Launchpad](https://help.sap.com/docs/ai-launchpad/sap-ai-launchpad-user-guide/models-and-scenarios-in-generative-ai-hub-fef463b24bff4f44a33e98bb1e4f3148#models).

```

#### Add Embeddings to Your CDS Model
Use the `cds.Vector` type in your CDS model to store embeddings. Set the dimension to match your embedding model (for example, 1536 for OpenAI *text-embedding-3-small*).

   ```cds
   entity Books : cuid {
     title       : String(111);
     description : LargeString;

>      embedding   : Vector(1536); // adjust dimensions to embedding model

Author: The embedding dimensions are a property of the chosen embedding model and documented by the model provider (usually also in the [mteb leaderboard](https://huggingface.co/spaces/mteb/leaderboard)).

For example the `gemini-embedding-001` has up to 3072 dimensions and `text-embedding-3-small` has 1536.

I'll try to make this clearer from the description.

>      embedding   : Vector(1536); // adjust dimensions to embedding model

Author: Added a section "Choose an Embedding Model"

### Vector Embeddings { #vector-embeddings }

Vector embeddings let you add semantic search, recommendations, and generative AI features to your CAP application. Embeddings are numeric arrays that represent the meaning of unstructured data (text, images, etc.), making it possible to compare and search for items that are semantically related to each other or a user query.

#### Choose an Embedding Model

Choose an embedding model that fits your use case and data (for example english or multilingual text). The model determines the number of dimensions of the resulting output vector. Check the documentation of the respective embedding model for details.

Use the [SAP Generative AI Hub](https://community.sap.com/t5/technology-blogs-by-sap/how-sap-s-generative-ai-hub-facilitates-embedded-trustworthy-and-reliable/ba-p/13596153) for unified consumption of embedding models and LLMs across different vendors and open source models. Check for available models on the [SAP AI Launchpad](https://help.sap.com/docs/ai-launchpad/sap-ai-launchpad-user-guide/models-and-scenarios-in-generative-ai-hub-fef463b24bff4f44a33e98bb1e4f3148#models).

#### Add Embeddings to Your CDS Model

> Use the `cds.Vector` type in your CDS model to store embeddings on SAP HANA Cloud. Set the dimension to match your embedding model (for example, 1536 for OpenAI *text-embedding-3-small*).

Reviewer: ```suggestion
Use the `cds.Vector` type in your CDS model to store embeddings on SAP HANA Cloud. Set the dimension to match your embedding model (for example, 1536 embedding dimensions for OpenAI *text-embedding-3-small*).
```

   ```cds
   entity Books : cuid {
     title       : String(111);
     description : LargeString;
     embedding   : Vector(1536); // adjust dimensions to embedding model
   }
   ```

#### Generate Embeddings
Use an embedding model to convert your data (for example, book descriptions) into vectors. The [SAP Cloud SDK for AI](https://sap.github.io/ai-sdk/) makes it easy to call SAP AI Core services to generate these embeddings.

:::details Example using SAP Cloud SDK for AI
```Java
var aiClient = OpenAiClient.forModel(OpenAiModel.TEXT_EMBEDDING_3_SMALL);
var response = aiClient.embedding(
   new OpenAiEmbeddingRequest(List.of(book.getDescription())));
book.setEmbedding(CdsVector.of(response.getEmbeddingVectors().get(0)));
```

> :::

Reviewer: @MattSchur Do you know whom we could ask from the Node.js side to provide an example?

> :::

Author: ```suggestion
```js [Node.js]
const response = await new AzureOpenAiEmbeddingClient(
 'text-embedding-3-small'
).run({
 input: bookDescription
});
const embedding = response.getEmbedding();
/```
:::
```
@MatKuhr is this correct? In CAP Node.js the vector embedding needs to be a string '[0.3,0.7,0.1,...]'

Reviewer 2: `response.getEmbedding()` returns `number[]`. Why would a string be expected? That seems unexpected to me.. Especially if its a string representation of an array..

Author: @David-Kunz is the Node runtime still expecting the embedding vector as String?

   ```cds
   entity Books : cuid {
     title       : String(111);
     description : LargeString;
     embedding   : Vector(1536); // adjust dimensions to embedding model
   }
   ```

2. **Generate Embeddings:** Use an embedding model to convert your data (for example, book descriptions) into vectors. The [SAP Cloud SDK for AI](https://sap.github.io/ai-sdk/) makes it easy to call SAP AI Core services to generate these embeddings.

   <div class="impl java">

   ```Java
   var aiClient = OpenAiClient.forModel(OpenAiModel.TEXT_EMBEDDING_3_SMALL); // [!code focus]

   var params = new OpenAiEmbeddingParameters();
   params.setInputType("document");
   params.setInput(book.getDescription()); // [!code focus:4]

   book.setEmbedding(CdsVector.of(
     aiClient.embedding(params).getData().get(0).getEmbedding()));
   ```
   </div>

3. **Query for Similarity:** At runtime, use SAP HANA’s built-in vector functions to search for similar items. For example, find books with embeddings similar to a user question:

   <div class="impl java">

   ```Java
   // Compute embedding for user question
   params.setInputType("query");

>    params.setInput("How to use vector embeddings in CAP?");

Reviewer 2: same adjustment as above

   ```cds
   entity Books : cuid {
     title       : String(111);
     description : LargeString;
     embedding   : Vector(1536); // adjust dimensions to embedding model
   }
   ```

2. **Generate Embeddings:** Use an embedding model to convert your data (for example, book descriptions) into vectors. The [SAP Cloud SDK for AI](https://sap.github.io/ai-sdk/) makes it easy to call SAP AI Core services to generate these embeddings.

   <div class="impl java">

   ```Java
   var aiClient = OpenAiClient.forModel(OpenAiModel.TEXT_EMBEDDING_3_SMALL); // [!code focus]

   var request = new OpenAiEmbeddingRequest(List.of(book.getDescription()));

   book.setEmbedding(CdsVector.of(
     aiClient.embedding(request).getEmbeddings().get(0)));
   ```
   </div>

3. **Query for Similarity:** At runtime, use SAP HANA’s built-in vector functions to search for similar items. For example, find books with embeddings similar to a user question:

   <div class="impl java">

   ```Java
   // Compute embedding for user question

>    params.setInputType("query");
>    params.setInput("How to use vector embeddings in CAP?");

Author: ```suggestion
   var request = new OpenAiEmbeddingRequest(List.of("How to use vector embeddings in CAP?"));
```
@MatKuhr should we set inputType = query?

   CdsVector userQuestion = CdsVector.of(
    aiClient.embedding(params).getData().get(0).getEmbedding());

   // Compute similarity between user question and book embeddings
   var similarity = CQL.cosineSimilarity( // computed on SAP HANA
     CQL.get(Books.EMBEDDING), userQuestion);

   // Find Books related to user question ordered by similarity
   hana.run(Select.from(BOOKS).limit(10)
			.columns(b -> b.ID(), b -> b.title(),
					 b -> similarity.as("similarity"))
			.orderBy(b -> b.get("similarity").desc()));
   ```
   </div>

   <div class="impl node">

   ```js
const response = await new AzureOpenAiEmbeddingClient(
    'text-embedding-3-small'
  ).run({
    input: 'Hello, world!'
  });

> const embedding = response.getEmbedding();

Author: ```suggestion
const questionEmbedding = response.getEmbedding();
```
